accepted/2020/diagnostics/activity-improvements-2.md

```diff
@@ -0,0 +1,144 @@
+        public ActivityKind Kind { get; set; } = ActivityKind.Internal;
+    }
+```
+
+## ActivityAttributesCollection
+
+OT has the concept of [Attributes](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#set-attributes) which is a list of key-value pairs mapping a string to some value. The value can be of different types (e.g. numeric, bool, string, arrays...etc.). 
+
+> ```json
+>      'x-forwarded-for': 'foo.bar'
+>      'timeout': 5
+> ```
+
+The attributes are used in multiple places in the OT (e.g. in Span, Event, and Link). Activity support Tags which similar concept but Tags is limited to string value types and not supporting any other types `IEnumerable<KeyValuePair<string, string?>> Tags`.  We'll need to extend the Tags concept in our APIs to support the OT scenarios. We'll do that by introducing the ActivityAttributesCollection and extending Activity class to support adding different types of attribute values.
+
+ActivityAttributesCollection is needed to ensure the OT specs behavior is implemented and avoid API users to use different type of collections (e.g. Dictionary<>, List<>...etc.) which not conforming with the OT specs. The OT specs require the attributes always stored in order and define behavior when adding values for keys which already existed in the list, it will override the old value. In addition it define the behavior what should happen when using null values.
+
+```c#
+    public class ActivityAttributesCollection : IReadOnlyCollection<KeyValuePair<string, object>>
+    {
+        public ActivityAttributesCollection() { }
+        public ActivityAttributesCollection(IEnumerable<KeyValuePair<string, object>>) { }
+
+        // If the key was not set before, it will get added to the collections with the input value.
+        // If the key was set before, the mapped value of this key will be updated with the new input value.
+        public void Set(string key, int value) { }
+
+        // interfaces implementation
+        public int Count { get; }
+        public IEnumerator<KeyValuePair<string, object>> GetEnumerator() { ... }
+        IEnumerator IEnumerable.GetEnumerator() { ... }
+    }
+```
+
+## Activity Attributes
+
+As mentioned earlier, Activity has Tags property which returns the list of the activity tags. Tags are a list of `KeyValuePair<string, string>`. Tags are proven to be not enough for OT scenarios as we need to support more value types (numeric, bool, arrays...etc.). Activity today has the method `public Activity AddTag(string key, string? value)`. The proposal here is to add `SetAttribute()` which can set attributes value for a string key. 
+
+In general, Attributes will be considered a superset of Tags. 
+
+```C#
+    public partial class Activity : IDisposable
+    {
+        public void SetAttribute(string key, object value) { }
+        
+        public ActivityAttributesCollection Attributes { get; }
```

Comment:
Can you confirm which of the following would be the recommended usage pattern?

`activity.Attributes.Set("key", "value")`

or

`activity.SetAttribute("key", "value");`

Reply:
This is raising a good point. Usaers should use 
```C#
activity.SetAttribute("key", "value");
```

and `Activity.Attributes` property is always going to return a new Collection instance (like what the Tags do today). If we think this can be confusing, we can return IEnumerable<...> instead of the collection. This will be less confusing. 

@cijothomas @noahfalk what you think?

Reply:
`activity.Tags` return `IEnumerable` only. I think making it same for `activity.Attributes` too is the best option. So the only way to set attribute is by using `activity.SetAttribute`.

Reply:
I am fine with returning IEnumerable (in Activity, ActivityEvent, and ActivitLink)